studio/src/components/CodeMirrorEditor/CodeMirrorInput.tsx

```diff
@@ -0,0 +1,211 @@
+  searchKeymap: false,
+  historyKeymap: false,
+  foldKeymap: false,
+  completionKeymap: false,
+  lintKeymap: false,
+};
+
+// An extension that does nothing to make conditional extensions easier,
+// e.g., `isFocused ? noopExtension : inputTrucateExtension`
+const noopExtension = EditorView.theme({});
+
+// Apply styles for the editor to make it look like one of our inputs
+const inputBaseStylesExtension = EditorView.theme({
+  ".cm-placeholder": {
+    color: "hsl(var(--muted-foreground))",
+    // HACK - Using tailwind `text-sm` values
+    fontSize: "0.875rem",
+    lineHeight: "1.25rem",
+  },
+  ".cm-scroller": {
+    overflow: "hidden !important",
+    maxWidth: "100% !important",
+  },
+});
+
+// Extension to truncate text that overflows with an ellipsis
+// We only want to use this when the editor is *not* focused
+const inputTrucateExtension = EditorView.theme({
+  ".cm-content": {
+    whiteSpace: "nowrap",
+    overflow: "hidden",
+    textOverflow: "ellipsis",
+    width: "100%",
+  },
+  ".cm-line": {
+    textOverflow: "ellipsis",
+    overflow: "hidden",
+    width: "100%",
+  },
+});
+
+// Extension for special styling in readonly mode, which makes the editor's cursor transparent,
+// but uses a pointer cursor when the user hovers over the input,
+// to indicate that you can click on it to expand it (in case the path parameter is truncated).
+// As of writing, this is only used for the path parameter *key* input.
+const readonlyExtension = EditorView.theme({
+  ".cm-cursor": {
+    borderLeftColor: "transparent !important",
+  },
+  ".cm-content": {
+    cursor: "pointer !important",
+  },
+});
+
+// Hide the gutter on the code mirror instance
+const hiddenGutterExtension = gutter({ class: "hidden border-none" });
+
+type CodeMirrorEditorProps = {
+  readOnly?: boolean;
+  width?: string;
+  value?: string;
+  onChange: (value?: string) => void;
+  placeholder?: string;
+  onSubmit?: () => void;
+};
+
+const preventNewlineInFirefox = keymap.of([
+  {
+    key: "Enter",
+    run: (_view) => {
+      // Do nothingggg and don't bubble up the event
+      return true;
+    },
+  },
+]);
+
+// Extension that blurs the editor when the user presses "Escape"
+const escapeKeymap = keymap.of([
+  {
+    key: "Escape",
+    run: (view) => {
+      view.contentDOM.blur();
+      return true;
+    },
+  },
+]);
+
+const submitKeymap = (onSubmit: (() => void) | undefined) =>
+  keymap.of([
+    {
+      key: "Mod-Enter",
+      run: (_view) => {
+        if (onSubmit) {
+          onSubmit();
+        }
+        return true;
+      },
+    },
+  ]);
+
+export function CodeMirrorInput(props: CodeMirrorEditorProps) {
+  const { value, onChange, placeholder, width, readOnly, onSubmit } = props;
+
+  const [isFocused, setIsFocused] = useState(false);
+
+  const style = useMemo(
+    () => ({
+      // if specified, the width should be a fixed width, otherwise we take up full width
+      width: width ?? "100%",
+      height: "auto",
```

Comment:
Bonus points if the static styling is done via the tailwind classes (which is also done)  and even more points if isFocused can be removed (something like:`focus-within:overlfow-auto`)

Reply:
good point — this style object was a holdover from the first attempts at writing a custom component. we can just set the width with a tailwind class that's passed as a prop! refactoring

Reply:
focus might be harder though, since the parent div needs to be aware of focus state, but i will try

Reply:
yeah so focus state needed to remain in order to conditionally load a codemirror extension that helps with truncation. but the styles have been refactored to rely on focus-within!